x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlTestCase.java

```diff
@@ -1183,4 +1204,207 @@ public static void assertResponse(Map<String, Object> expected, Map<String, Obje
+        final int fetchCount = randomIntBetween(1, 9);
+        bulkLoadTestData(fetchSize * fetchCount); // NB: product needs to stay below 100, for txt format tests
+
+        String format = randomFrom(acceptMap.keySet());
+        String mode = randomMode();
+        String cursor = null;
+        for (int i = 0; i <= fetchCount; i++) { // the last iteration (the equality in `<=`) checks on no-cursor & no-results
+            // start the query
+            RequestObjectBuilder builder = (hasText(cursor) ? cursor(cursor) : query("SELECT text, number FROM test")).fetchSize(fetchSize)
+                .waitForCompletionTimeout("0d") // don't wait at all
+                .keepOnCompletion(true)
+                .keepAlive("1d") // keep "forever"
+                .mode(mode)
+                .binaryFormat(false); // prevent JDBC mode to (ignore `format` and) enforce CBOR
+            Response response = runSqlAsTextWithFormat(builder, format);
+
+            Character csvDelimiter = ',';
+
+            assertEquals(200, response.getStatusLine().getStatusCode());
+            assertEquals(response.getHeader(HEADER_NAME_ASYNC_PARTIAL), response.getHeader(HEADER_NAME_ASYNC_RUNNING));
+            String asyncId = response.getHeader(HEADER_NAME_ASYNC_ID);
+            assertTrue(hasText(asyncId));
+
+            // it happens though rarely that the whole response comes through despite the given 0-wait
+            if (response.getHeader(HEADER_NAME_ASYNC_PARTIAL).equals("true")) {
+
+                // potentially wait for it to complete
+                boolean pollForCompletion = randomBoolean();
+                if (pollForCompletion) {
+                    Request request = new Request("GET", SQL_ASYNC_STATUS_REST_ENDPOINT + asyncId);
+                    Map<String, Object> asyncStatus = null;
+                    long millis = 1;
+                    for (boolean isRunning = true; isRunning; Thread.sleep(millis *= 2)) {
+                        asyncStatus = toMap(client().performRequest(request), null);
+                        isRunning = (boolean) asyncStatus.get(IS_RUNNING_NAME);
+                    }
+                    assertEquals(200, (int) asyncStatus.get("completion_status"));
+                    assertEquals(asyncStatus.get(IS_RUNNING_NAME), asyncStatus.get(IS_PARTIAL_NAME));
+                    assertEquals(asyncId, asyncStatus.get(ID_NAME));
+                }
+
+                // fetch the results (potentially waiting now to complete)
+                Request request = new Request("GET", SQL_ASYNC_REST_ENDPOINT + asyncId);
+                if (pollForCompletion == false) {
+                    request.addParameter(WAIT_FOR_COMPLETION_TIMEOUT_NAME, "1d");
+                }
+                if (randomBoolean()) {
+                    request.addParameter(URL_PARAM_FORMAT, format);
+                    if (format.equals("csv")) {
+                        csvDelimiter = ';';
```

Comment:
If I understand this well, you use the default delimiter (`,`) but depending on a `randomBoolean` you change that (and is set explicitly) to `;`. I would move `Character csvDelimiter = ',';` closer to this `randomBoolean`. If I wouldn't have put the code in the IDE and look at what happens with the `csvDelimiter` it would have been more difficult to see what the logic is.

Reply:
The `randomBoolean()` splits between passing the `format` as a URL parameter _or_ HTTP header. Then, if the chosen branch is to use the parameter _and_ the format - also randomly chosen - is `csv` then the splitting character is changed.
The `randomBoolean()` isn't solely responsible for the delimiter char, so not sure how to make it clearer, but if you have a suggestion I'll apply it.